### PR TITLE
Integrate Angle Field into the MLP for line classification

### PR DESCRIPTION
**Pull request by hardik01shah**
- Add code for data generation for MLP training in `gluefactory/models/mlp/generate_with_AF.py`.
- Add MLP architecture and training code in `gluefactory/models/mlp/train_with_AF.py`.
- Fix matching_score implementation and calling in `gluefactory/utils/desc_evaluation.py` and `gluefactory/eval/hpatches_extended.py` respectively.

### CODE REVIEW

**Comment by hardik01shah** [comment]
# **Major Refactor**

## Adding MLP as a Gluefactory model (POLD2_MLP) that derives `BaseModel` in `gluefactory/models/lines/pold2_mlp.py`
Use the following command to train the MLP
```sh
python -m gluefactory.train pold2_mlp_test --conf gluefactory/configs/pold2_mlp_train.yaml
```

Use the following command to visualize the confusion matrix on the validation dataset.
```sh
python -m gluefactory.models.lines.pold2_mlp \
        --conf gluefactory/configs/pold2_mlp_train.yaml \
        --weights outputs/training/pold2_mlp_test/checkpoint_best.tar
```
## Add a Gluefactory dataset in `gluefactory/datasets/pold2_mlp_dataset.py` for dataloading and generation of data for MLP training
Use the following command to test the dataloader
```sh
python -m gluefactory.datasets.pold2_mlp_dataset --conf gluefactory/configs/pold2_mlp_dataloader_test.yaml
```
## Update `gluefactory/models/lines/pold2_extractor.py`

- Use the MLP model defined in `gluefactory/models/lines/pold2_mlp.py` as the strong filter
- Clean up - remove unnecessary functions, parameters
- Make `LineExtractor` derive `BaseModel` making it callable
- Add all tunable parameters as config parameters, remove inline assignment

Use the following command to test `LineExtractor` on the oxford-paris (POLD2) dataset:
```sh
python -m gluefactory.models.lines.pold2_extractor --conf gluefactory/configs/pold2_line_extractor_test.yaml --show
```
## TODO

- NMS for lines `merge_lines_torch` is called outside of `LineExtractor`. Add a parameter in the config for NMS, and merge lines, and return the line endpoints, along with keypoint indices. **Will affect all instances of POLD2 which expects 'lines' to be keypoint indices and not line endpoints.**
- `LineExtractor` usage across the repository must be updated.

File: gluefactory/configs/pold2_mlp_dataloader_test.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+data_dir: "pold2_mlp_dataset"
```
Comment (FelixMoeller3) [inline gluefactory/configs/pold2_mlp_dataloader_test.yaml:1]: Is there a reason why the data_dir is just the directory itself whereas the weights are given given as relative paths?

File: gluefactory/datasets/homographies_point_line.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def plot_predictions(pred, data):
     kp_0 = data["view0"]["cache"]["keypoints"][0].cpu().numpy()
     df_0 = data["view0"]["cache"]["df"][0].cpu().numpy()
     angle_0 = data["view0"]["cache"]["line_level"][0].cpu().numpy()
-    angle_0 = np.arctan2(angle_0[1, :], angle_0[0, :])
+    angle_0 = np.arctan2(angle_0[1], angle_0[0])
```
Comment (FelixMoeller3) [inline gluefactory/datasets/homographies_point_line.py:59]: Why did the indices change here? (just asking to understand :D )

**Review by FelixMoeller3 — APPROVED** [review]
Looks mostly good to me. The comments are mainly for my understanding :)

File: gluefactory/datasets/homographies_point_line.py (hunk 2 above)
Comment (hardik01shah) [reply gluefactory/datasets/homographies_point_line.py:59]: the indexing for next dimension using ' : ' was basically pointless and did not make a difference. So I removed it to keep it the same as line 66.

File: gluefactory/configs/pold2_mlp_dataloader_test.yaml (hunk 1 above)
Comment (hardik01shah) [reply gluefactory/configs/pold2_mlp_dataloader_test.yaml:1]: Yeah, I thought over the course of testing the MLP we would keep exchanging weight files so better keep those as absolute paths and the data hopefully should not change a lot so kept it relative to the DATA_DIR of settings.py.

**Pull request merged**